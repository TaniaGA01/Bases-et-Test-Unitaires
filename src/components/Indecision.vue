<template>
  <img v-if="img" :src="img" />
  <!-- los dos puntos reemplazan v-bind -->
  <div class="bg-dark"></div>
  <div class="indecision-container">
    <input type="text" placeholder="Hazme une pregunta" v-model="newQuestion" />
    <p>Recuerda terminar con un punto de interrogación (?)</p>
    <div v-if="isValideQuestion">
      <h2>{{ newQuestion }}</h2>
      <h1>{{ answer }}</h1>
    </div>
  </div>
</template>

<script>
export default {
  data() {
    return {
      newQuestion: null,
      answer: null, // este answer como tal viene del API https://yesno.wtf/api
      img: null,
      isValideQuestion: false,
    };
  },
  methods: {
    async getAnswer() {

      try {

        this.answer = "Attendez...";
        const { answer, image } = await fetch('https://yesno.wtf/api').then( r => r.json() )

        this.answer = answer === 'yes' ? 'Oui !' : 'Non !'// la primera answer es el return de data la segunda es la const
        this.img = image // la primera img es el return de data la segunda image es la const
        
      } catch (error) {

        console.log('Indecision Component:', error)
        this.answer = 'Impossible de charger de l\'API'
        this.img = null

      }
    },
  },
  watch: {
    newQuestion(value) {
      this.isValideQuestion = false;
      console.log({ value });
      if (!value.includes("?")) return;
      console.log({ value });
      this.getAnswer();

      this.isValideQuestion = true;
    },
  },
};
</script>

<style scoped>
/* Scoped : css dedicado únicamente a este componente */

img,
.bg-dark {
  height: 100vh;
  left: 0px;
  max-height: 100%;
  max-width: 100%;
  position: fixed;
  top: 0px;
  width: 100vw;
}

.bg-dark {
  background-color: rgba(0, 0, 0, 0.4);
}

.indecision-container {
  position: relative;
  z-index: 99;
}

input {
  width: 250px;
  padding: 10px 15px;
  border-radius: 5px;
  border: none;
}
input:focus {
  outline: none;
}

p {
  color: white;
  font-size: 20px;
  margin-top: 0px;
}

h1,
h2 {
  color: white;
}

h2 {
  margin-top: 150px;
}
</style>