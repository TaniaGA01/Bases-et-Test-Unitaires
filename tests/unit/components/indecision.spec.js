/*  
Pour la maj il faut utiliser la commande 'yarn test:unit -u indecision'
*/
import { shallowMount } from "@vue/test-utils";
import Indecision from '@/components/Indecision';

describe('Indecision Component', () => {

    let wrapper;
    let clgSpy; // Les spy's sont des espions qui attends des événements

    // Mock : résoudre erreurs dues aux espions sur le fetch (ligne 29 Indecision.vue) :
    global.fetch = jest.fn( () => Promise.resolve({ 
        json: () => Promise.resolve({ // les propriétés suivantes viennent de https://yesno.wtf/#api
            answer: 'yes',
            forced: false,
            image: "https://yesno.wtf/assets/yes/2.gif"
        })
    }) )


    beforeEach(() => {

        wrapper = shallowMount(Indecision)
        clgSpy = jest.spyOn(console, 'log')// fonction espion

        jest.clearAllMocks() // netoie le cache pour ne pas avoir des repetition donc des erreurs

    });

     // TEST 1 : SnapShot, une copie du composant sera créée dans un fichier __snapshot__
    test('Avoir match avec le snapshot', () => {

        expect(wrapper.html()).toMatchSnapshot()

    })
     // TEST 2 : Test dans l'input
    test('Écrire dans l\'input sans le symbole ? ne doit rien envoyer (console.log)', async () => {

        const getAnswerSpy = jest.spyOn(wrapper.vm, 'getAnswer')// vm --> voir doc https://v1.test-utils.vuejs.org/api/wrapper/

        const input = wrapper.find('input')
        await input.setValue('Hello World')

        expect(clgSpy).toHaveBeenCalledTimes(1)
        expect(getAnswerSpy).not.toHaveBeenCalled()

    })
     // TEST 3 : Test dans l'input avec le symbole ?
    test('Écrire dans l\'input avec le symbole ? doit activer le getAnswer', async() => {

        const getAnswerSpy = jest.spyOn(wrapper.vm, 'getAnswer')// vm --> voir doc https://v1.test-utils.vuejs.org/api/wrapper/

        const input = wrapper.find('input')
        await input.setValue('Hello World ?')

        expect(clgSpy).toHaveBeenCalledTimes(2)
        expect(getAnswerSpy).toHaveBeenCalled()


    })
     // TEST 4 : Test getAnswer
    test('Test sur le getAnswer', async() => {

        await wrapper.vm.getAnswer()

        const img = wrapper.find('img')

        expect(img.exists()).toBeTruthy()
        expect(wrapper.vm.img).toBe('https://yesno.wtf/assets/yes/2.gif')
        expect(wrapper.vm.answer).toBe('Oui !')

    })
    // TEST 5 : Test getAnswer + faille avec l'API
    test('Test sur le getAnswer avec faille de l\'API', async() => {

        //simuler la faille du fetch :
        fetch.mockImplementationOnce(() => Promise.reject('API est down'))
        
        await wrapper.vm.getAnswer()

        const img = wrapper.find('img')

        expect(img.exists()).toBeFalsy()
        expect(wrapper.vm.answer).toBe('Impossible de charger de l\'API')
    })


})