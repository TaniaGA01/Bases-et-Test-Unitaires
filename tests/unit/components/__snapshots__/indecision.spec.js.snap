// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Indecision Component Avoir match avec le snapshot 1`] = `
<!--v-if-->
<!-- los dos puntos reemplazan v-bind -->
<div class="bg-dark"></div>
<div class="indecision-container"><input type="text" placeholder="Hazme une pregunta">
  <p>Recuerda terminar con un punto de interrogaci√≥n (?)</p>
  <!--v-if-->
</div>
`;
